File: geohash/geohash.py
class GeoHash:
    # @param {double} latitude, longitude a location coordinate pair
    # @param {int} precision an integer between 1 to 12
    # @return {string} a base32 string
    def encode(self, latitude, longitude, precision):
        # Write your code here
        base32 = "0123456789bcdefghjkmnpqrstuvwxyz"
        c_lo = (precision * 5 + 1) / 2
        c_la = precision * 5 / 2;
        b = []
        bo = self.get_b(longitude, -180, 180, c_lo)
        ba = self.get_b(latitude, -90, 90, c_la)
        for i in xrange(c_la):
            b.append(bo[i])
            b.append(ba[i])
        if c_lo > c_la:
            b.append(bo[-1])
        ans = ""
        for i in xrange(precision):
            id = self.get_id(5 * i, 5 * i + 4, b)
            ans += base32[id]
        return ans
    
    def get_b(self, x, l, r, n):
        b = []
        for i in xrange(n):
            m = (l + r) * 0.5
            if x <= m:
                r = m
                b.append(0)
            else:
                l = m
                b.append(1)
        return b
    
    def get_id(self, l, r, b):
        ans = 0
        for i in range(l, r + 1):
            ans = ans * 2 + b[i]
        return ans

File: geohash/geohash.cc
class GeoHash {
    public:
    /**
     * @param latitude, longitude a location coordinate pair
     * @param precision an integer between 1 to 12
     * @return a base32 string
     */
    string encode(double latitude, double longitude, int precision) {
        // Write your code here
        int c_lo, c_la;
        c_lo = (5 * precision + 1) / 2;
        c_la = 5 * precision / 2;
        get_bin(longitude, -180, 180, c_lo, 0);
        get_bin(latitude, -90, 90, c_la, 1);
        string ans;
        for (int i = 0; i < precision; ++i) {
            int id = get_id(i * 5, i * 5 + 4);
            ans += base32[id];
        }
        return ans;
    }
    
    void get_bin(double x, double l, double r, int n, int a) {
        double m;
        for (int i = 0; i < n; ++i) {
            m = (l + r) / 2;
            if (x <= m) {
                bin[i * 2 + a] = 0;
                r = m;
            } else {
                bin[i * 2 + a] = 1;
                l = m;
            }
        }
    }
    
    int get_id(int l, int r) {
        int id = 0;
        for (int i = l; i <= r; ++i) {
            id = id * 2 + bin[i];
        }
        return id;
    }
    
    private:
    string base32 = "0123456789bcdefghjkmnpqrstuvwxyz";
    int bin[60];
};
